Hsearch + quarkus blog post (#141)
hibernate/in.relation.to: posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc

= Hibernate Search in Quarkus
Yoann Rodière
:awestruct-tags: [ "Hibernate Search", "Elasticsearch", "Quarkus" ]
:awestruct-layout: blog-post
:hsearchDocUrl: https://docs.jboss.org/hibernate/search/6.0/reference/en-US/html_single/
:demoRepoUrl: https://github.com/hibernate/hibernate-demos/tree/master/hibernate-search/hsearch-quarkus
---

Hibernate libraries can be used with many frameworks,
and we're striving to make sure that everyone gets their fair share of Hibernate goodness.
So with the recent announce of https://quarkus.io/blog/announcing-quarkus-1-0/[the first release candidate of Quarkus],
the "supersonic, subatomic" Java stack,
it's worth mentioning that Hibernate ORM, Search and Validator are already included in Quarkus.

Hibernate Search in particular, the library adding the power of full-text search to your ORM-based application,
is included in its most recent version (6.0.0.Beta2), which adds first-class compatibility with Elasticsearch.

Let's take this Quarkus release as an opportunity to have a closer look at Hibernate Search 6,
and how it can be used in Quarkus.

+++<!-- more -->+++

== Context

=== What's Quarkus?

> https://quarkus.io/[Quarkus] is a Java stack for developing web applications.

Reviewer: It's more than web-apps, e.g. also stream processing or CLI tools eventually. It's a generic app dev stack.

Author: That's not what https://quarkus.io/ seems to focus on, but alright, I'll change it.

It integrates many Java libraries and provides tooling to ease development.
Quarkus-based applications can run in a traditional HotSpot setup,
but also (via https://www.graalvm.org/[GraalVM])

> as native binaries with a tiny memory footprint and impressively fast boot time.

Reviewer 2: I think I would separate the two sentences: it's important that people understand that Quarkus have outstanding benefits with HotSpot too.

The latter is what really sets Quarkus apart:

> it means that Quarkus-based applications are especially well-suited for running in a container.

Reviewer 2: At some point, we need to mention the live coding experience too, because that's the one thing that people really like and makes a real difference.

http://hibernate.org/orm/[Hibernate ORM] to handle persistence to a database,
and http://hibernate.org/search/[Hibernate Search] to add full-text search features to the application.

The application will be backed by https://www.postgresql.org/[PostgreSQL] to store normalized, relational data,
and https://www.elastic.co/products/elasticsearch[Elasticsearch] to index part of this data in a de-normalized,
easy-to-search fashion.

== Code

You can find the code of the resulting REST application
in the `hibernate/hibernate-demos` repository on GitHub:
link:{demoRepoUrl}[].

== Prerequisites

The application described in the following sections was built using the following tools:

* A shell
* OpenJDK 8 (OpenJDK 13 works just as well).
You can get binaries from https://www.oracle.com/technetwork/java/javase/downloads/index.html[Oracle],
or from the https://adoptopenjdk.net/[AdoptOpenJDK project], among others.
* https://maven.apache.org/download.cgi[Apache Maven] 3.6.2.
* https://www.docker.com/[Docker] and in particular https://docs.docker.com/compose/[Docker-compose]
to run PostgreSQL and Elasticsearch.

> GraalVM is necessary in order to build a native image of the application;

Reviewer: It can be done via an in-container build, too, which I'd recommend for easier set-up.

Author: Right, I did that but forgot to update this line. Will do.

see https://quarkus.io/get-started/ for details.

== Initial CRUD application

Before we add full-text search to our application,
we need some data to work with.
This section will explain how to create a basic CRUD application with Quarkus.

If you are already familiar with this, just skip through this section and go right to <<adding-hibernate-search>>.

=== Initializing the project

Quarkus provides a Maven plugin to initialize the layout of a new project.
Just run this:

[source]
----

> $ mvn io.quarkus:quarkus-maven-plugin:0.28.0:create \

Reviewer 2: Let's use 1.0.0.CR1?

Author: I did, and updated all the examples. I just forgot to update this one.

    └── test
        └── java
            └── org
                └── hibernate
                    └── demos
                        └── quarkus
                            ├── ClientResourceTest.java
                            └── NativeClientResourceIT.java

17 directories, 10 files
----

=== Tests

The tests created by Quarkus are fine,
and in a real-world application we should update them as we add more features to our applications.

However, for the sake of brevity, we will not tackle tests here.
Let's just delete them:

[source]
----
$ rm -rf src/test
----

=== Environment

The easiest way to reliably run PostgreSQL and Elasticsearch in your development environment is probably to use docker.

A docker-compose configuration file is available
link:{demoRepoUrl}/environment-stack.yml[here].
It includes a cluster of two Elasticsearch nodes and a PostgreSQL instance.

We can start it like this:

[source]
----
$ docker-compose -f environment-stack.yml -p hsearch-quarkus-env up
----

And stop it, removing all docker volumes, like this:

[source]
----
$ docker-compose -f environment-stack.yml -p hsearch-quarkus-env down -v
----

=== Configuration properties

> The configuration of database access will go into Quarkus' main configuration file: `application.properties`:

Reviewer 2: Maybe for the first occurrence, you should git the full path?

[source]
----
quarkus.ssl.native=false <1>

quarkus.datasource.url=jdbc:postgresql://${POSTGRES_HOST}/${POSTGRES_DB} <2>
quarkus.datasource.driver=org.postgresql.Driver
quarkus.datasource.username=${POSTGRES_USER}
quarkus.datasource.password=${POSTGRES_PASSWORD}
%dev.POSTGRES_HOST=localhost <3>
%dev.POSTGRES_DB=hsearch_demo
%dev.POSTGRES_USER=hsearch_demo

> %dev.POSTGRES_PASSWORD=hsearch_demo

Reviewer 2: This is a bit weird.

Author: Ok, I'll need a hint. What is weird?

Author: As discussed, the fact that I override environment variables from this file. I'll change it.

If it's not already done, let's start PostgreSQL:

[source]
----
$ docker-compose -f environment-stack.yml -p hsearch-quarkus-env up
----

Then let's compile and run the application in development mode:

[source]
----
$ ./mvnw clean compile quarkus:dev
[INFO] Scanning for projects...
[INFO]
[INFO] ----------------< org.hibernate.demos:hsearch-quarkus >-----------------
[INFO] Building hsearch-quarkus 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ hsearch-quarkus ---
[INFO] Deleting /home/yrodiere/workspaces/contributor-support/hibernate-demos/hibernate-search/hsearch-quarkus/target
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ hsearch-quarkus ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 3 resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ hsearch-quarkus ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 9 source files to /home/yrodiere/workspaces/contributor-support/hibernate-demos/hibernate-search/hsearch-quarkus/target/classes
[INFO]

> [INFO] --- quarkus-maven-plugin:1.0.0.CR1:dev (default-cli) @ hsearch-quarkus ---

Reviewer 2: See that's CR1 you're using!

2019-11-06 16:01:11,270 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (main) relation "client" does not exist, skipping
2019-11-06 16:01:11,271 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (main) SQL Warning Code: 0, SQLState: 00000
2019-11-06 16:01:11,271 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (main) table "businessmanager" does not exist, skipping
2019-11-06 16:01:11,272 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (main) SQL Warning Code: 0, SQLState: 00000
2019-11-06 16:01:11,272 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (main) table "client" does not exist, skipping
2019-11-06 16:01:11,272 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (main) SQL Warning Code: 0, SQLState: 00000
2019-11-06 16:01:11,273 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (main) sequence "hibernate_sequence" does not exist, skipping
2019-11-06 16:01:11,273 WARN  [io.agr.pool] (main) Datasource '<default>': JDBC resources leaked: 0 ResultSet(s) and 1 Statement(s)
2019-11-06 16:01:11,346 WARN  [io.agr.pool] (main) Datasource '<default>': JDBC resources leaked: 0 ResultSet(s) and 1 Statement(s)

> 2019-11-06 16:01:11,809 INFO  [io.quarkus] (main) Quarkus 1.0.0.CR1 started in 5.259s. Listening on: http://0.0.0.0:8080

Reviewer 2: 5 seconds? This sounds like a lot?

Author: It does. Stopping and recompiling and running it again once or twice will bring down the startup time to 2.5s, so I suspect it has something to do with PostgreSQL being slow for the first few commands, but I don't know what is wrong exactly.

2019-11-06 16:01:11,812 INFO  [io.quarkus] (main) Profile dev activated. Live Coding activated.
2019-11-06 16:01:11,813 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, hibernate-orm-panache, hibernate-search-elasticsearch, jdbc-postgresql, narayana-jta, resteasy, resteasy-jsonb]
----

We can call the REST service and check that the data is already there:

[source]
----
$ curl -X GET http://localhost:8080/client/2

{
    "assignedManager": {
        "email": "dschrute@dundermifflin.net",
        "id": 1,
        "name": "Dwight Schrute",
        "phone": "+1-202-555-0151"
    },
    "id": 2,
    "name": "Aperture Science Laboratories"
}
----

[[adding-hibernate-search]]
== Adding Hibernate Search

=== Dependencies

When we generated the project using Quarkus,
we told it we needed Hibernate Search.
Thus the dependencies are already present in the POM.

> If you are curious, the dependency looks like this:

Reviewer: Perhaps show the add-extension stuff instead, for the better effect instead of having it in there from the beginning?

Author: Good idea! Will do.

quarkus.hibernate-search.elasticsearch.index-defaults.lifecycle.strategy=create <4>
%dev.quarkus.hibernate-search.elasticsearch.index-defaults.lifecycle.strategy=drop-and-create <5>
%dev.quarkus.hibernate-search.elasticsearch.index-defaults.lifecycle.required-status=yellow <6>
----
<1> Hibernate Search needs to know the version of Elasticsearch it's going to connect to,
because different versions of Elasticsearch have different capabilities.
<2> Connection info is extracted from an environment variable.
This allows for easier deployment in cloud environments.
<3> In our development environment, we will always use the same connection info.
This sets the default value for the environment variable when running with the `dev` profile.
<4> By default, we will let Hibernate Search create the Elasticsearch schema on startup if it doesn't exist.
<5> In our development environment, we will drop and re-create the indexes on each startup (or on hot reload).
<6> In our development environment, we will allow the application to start
even if the indexes are in yellow status (not replicated).

=== Mapping

Hibernate Search is now aware of where to send indexed data,

> but it does not now what to send yet.

Reviewer 2: ```suggestion
but it does not know what to send yet.
```

> The definition of which parts of the entities needs to be indexed in Elasticsearch is called the _mapping_.

Reviewer 2: ```suggestion
The definition of which parts of the entities nees to be indexed in Elasticsearch is called the _mapping_.
```

The easiest way to map entities in Hibernate Search is using annotations:

[source,JAVA]
----
@Entity
@Indexed // <1>
public class Client extends PanacheEntity {

	@FullTextField(analyzer = "standard") // <2>
	public String name;

	@ManyToOne
	public BusinessManager assignedManager;

}
----
<1> Every entity we want to see mapped to an index needs to be annotated with `@Indexed`.
The index name, by default, will be the entity name (in this case `client`),
but that can be overridden using `@Indexed(index = "myindexname")`.
<2> By default, the document sent to the index for each entity is empty, which is not very useful.
New content is added by defining _fields_.
Here we define a field whose content will be extracted from the `name` property.

> It is full-text field, i.e. a text field which will be split into words upon indexing.

Reviewer 2: ```suggestion
It is a full-text field, i.e. a text field which will be split into words upon indexing.
```

Other types of fields exist, with link:{hsearchDocUrl}#mapper-orm-directfieldmapping-annotations[different annotations].
For now we're using the "standard" analyzer; we'll discuss this in more depth <<leveraging-full-text-search,further down>>.

=== Hot reload

We're now ready to start the application with Hibernate Search.

> Thanks to Quarkus' hot reload, if the application was already started with `quarkus:dev` when we performed the changes,

Reviewer 2: ```suggestion
Thanks to Quarkus' live coding feature, if the application was already started with `quarkus:dev` when we performed the changes,
```

      "value" : 1,
      "relation" : "eq"
    },
    "max_score" : 1.2300112,
    "hits" : [
      {
        "_index" : "client",
        "_type" : "_doc",
        "_id" : "2",
        "_score" : 1.2300112,
        "_source" : {
          "name" : "Aperture Science Laboratories"
        }
      }
    ]
  }
}
----

For a better development experience,
if the test dataset is small,
it is possible to trigger reindexing automatically on startup by adding this method to `ClientResource`:

[source,JAVA]
----
// ...
@Transactional
// ...
public class ClientResource {
	// ...

	@Transactional(TxType.NEVER)
	void reindexOnStart(@Observes StartupEvent event) throws InterruptedException {
		if ( "dev".equals( ProfileManager.getActiveProfile() ) ) {
			reindex();
		}
	}

	// ...
}
----

=== Automatic indexing

While mass indexing is convenient in some cases,
what's even more convenient is not having to care about indexing at all.
Hibernate Search provides what is called _automatic indexing_:
each time an entity is created, updated or deleted through a Hibernate ORM entity manager/session,
Hibernate Search will detect these changes and reindex the relevant entities as appropriate.

> Automatic is enabled by default, is completely transparent and requires no configuration.

Reviewer 2: ```suggestion
Automatic indexing is enabled by default, is completely transparent and requires no configuration.
```
?

and will automatically push the analyzer definitions to Elasticsearch when it creates the indexes.

We just need to implement an analyzer configurer:

[source,JAVA]
----
package org.hibernate.demos.quarkus.search;

import org.hibernate.search.backend.elasticsearch.analysis.ElasticsearchAnalysisConfigurationContext;
import org.hibernate.search.backend.elasticsearch.analysis.ElasticsearchAnalysisConfigurer;

public class ClientElasticsearchAnalysisConfigurer implements ElasticsearchAnalysisConfigurer {
	@Override
	public void configure(ElasticsearchAnalysisConfigurationContext context) {
		context.analyzer( "english" ).custom() // <1>
				.tokenizer( "standard" ) // <2>
				.tokenFilters( "lowercase", "stemmer_english", "asciifolding" ); // <3>
		context.tokenFilter( "stemmer_english" ) // <4>
				.type( "stemmer" )
				.param( "language", "english" );
	}
}
----
<1> Declare a custom analyzer named `english`.
<2> Set the tokenizer to `standard`, i.e. require that the analyzer generates words by splitting text on spaces, tabs, punctuation, etc.
<3> Apply three token filters to transform the extracted words: `lowercase` which turns the words to lowercase,

> `stemmer_english` which is a custom filter (see below), `asciifolding` which replaces accented characters with they ascii counterpart

Reviewer 2: ```suggestion
`stemmer_english` which is a custom filter (see below), `asciifolding` which replaces accented characters with their ascii counterpart
```

(`déjà-vu` => `deja-vu`).
<4> Declare a custom token filter named `stemmer_english`.
This token filter is a stemmer, meaning it will normalize the end of words (`laboratories` => `laboratory`),
and we configure it to handle the English language.

Then, we need to tell Hibernate Search to use our configurer by setting a configuration property in `application.properties`:

[source]
----
quarkus.hibernate-search.elasticsearch.analysis.configurer=org.hibernate.demos.quarkus.search.ClientElasticsearchAnalysisConfigurer

> ----

Reviewer 2: See how cool it would be to just have an annotation ;)

Author: Yep, it's called `@Named` but for some obscure reason it won't work in the  Quarkus Hibernate Search extension :)

And finally, we need to set the correct analyzer on our full-text field:

[source,JAVA]
----
@Entity
@Indexed
public class Client extends PanacheEntity {

	@FullTextField(analyzer = "english") // <1>
	public String name;

	@ManyToOne
	public BusinessManager assignedManager;

}
----
<1> Change the analyzer from `standard` to `english`.

After these changes, we need to restart the application and reindex the data.
Quarkus will do it automatically, so we can test the result of our changes right away:

[source]
----
$ curl -X GET 'http://localhost:8080/client/search/?terms=laboratory'

[
    {
        "assignedManager": {
            "email": "dschrute@dundermifflin.net",
            "id": 1,
            "name": "Dwight Schrute",
            "phone": "+1-202-555-0151"
        },
        "id": 2,
        "name": "Aperture Science Laboratories"
    }
]
----

> It worked: the test "laboratory" now matches the name "Aperture Science Laboratories".

Reviewer 2: ```suggestion
It worked: the text "laboratory" now matches the name "Aperture Science Laboratories".
```
?

2. When other properties of the business manager change (for example the phone number),
we do not need to reindex the assigned clients,
since these other properties are not indexed.

These two requirements would make manually reindexing entities significantly harder to implement efficiently:
the code would need to be aware of which of the business manager's properties are used when indexing a client,
it would need to keep track of which properties of the business manager's are actually changed,
and based on that would need to decide whether to load clients for reindexing or not.

Add a couple more associations like this to the `Client` entity or (worse) add a few levels of nesting,
and the simple boilerplate code will soon turn into a time sink.

Fortunately, Hibernate Search handles all this transparently.
In order to index the business manager's name as part of the client,
only two steps are necessary.

First, we will declare a field in the business manager:

[source,JAVA]
----
@Entity
public class BusinessManager extends PanacheEntity {

	@OneToMany(mappedBy = "assignedManager")
	public List<Client> assignedClients = new ArrayList<>();

> 	@FullTextField(analyzer = "english") // <1>

Reviewer 2: I would argue that this is not a good analyzer for names :).

Author: Fair enough, but I think after 1400 lines, I really should stop adding content to this post :) I'll add a comment about this near the end.

	public String name;

	public String email;

	public String phone;

}
----
<1> Define a full-text field whose content will be extracted from the `name` property.

[source,JAVA]
----
@Entity
@Indexed
public class Client extends PanacheEntity {

	@FullTextField(analyzer = "english")
	public String name;

	@ManyToOne
	@IndexedEmbedded // <1>
	public BusinessManager assignedManager;

}
----
<1> Define the assigned manager as "indexed-embedded" into the client,

> meaning all the index fields defined in the business manager will be embedded into the client upon indexing.

Reviewer 2: ```suggestion
meaning all the indexed fields defined in the business manager will be embedded into the client upon indexing.
```

> Simply put, a new field will appear in index documents generated for clients: `assignedManager.name`.

Reviewer 2: ```suggestion
Simply put, a new field will appear in indexed documents generated for clients: `assignedManager.name`.
```
Not sure about that one, maybe you really want to talk about the index.

we will simply use a https://quarkus.io/guides/building-native-image.html#creating-a-container-with-a-multi-stage-docker-build[multi-stage Docker build]
that will build our application in a container, then generate a container for our application.

Let's add a Dockerfile at `src/main/docker/Dockerfile.multistage`:

[source]
----
## Stage 1 : build with maven builder image with native capabilities
FROM quay.io/quarkus/centos-quarkus-maven:19.2.1 AS build
COPY src /usr/src/app/src
COPY pom.xml /usr/src/app
USER root
RUN chown -R quarkus /usr/src/app
USER quarkus
RUN mvn -f /usr/src/app/pom.xml -Pnative clean package

## Stage 2 : create the docker final image
FROM registry.access.redhat.com/ubi8/ubi-minimal
WORKDIR /work/
COPY --from=build /usr/src/app/target/*-runner /work/application
RUN chmod 775 /work
EXPOSE 8080
CMD ["./application", "-Dquarkus.http.host=0.0.0.0"]
----

Then let's build it (it will take some time, we're compiling a native binary here):

[source]
----
$ docker build -f src/main/docker/Dockerfile.multistage -t quarkus/hsearch-quarkus .
[... lots of logs ...]
Successfully tagged quarkus/hsearch-quarkus:latest
----

> The Docker image is now ready to be used.

Reviewer: container image

Let's start an environment if it's not already done:

[source]
----
$ docker-compose -f environment-stack.yml -p hsearch-quarkus-env up
----

And once everything is ready, let's start our application:

[source]
----
$ docker run --rm -it --network=host \
        -e POSTGRES_HOST=localhost \
        -e POSTGRES_DB=hsearch_demo \
        -e POSTGRES_USER=hsearch_demo \
        -e POSTGRES_PASSWORD=hsearch_demo \
        -e ES_HOSTS=http://localhost:9200 \
        quarkus/hsearch-quarkus

> 2019-11-07 10:33:15,520 INFO  [io.quarkus] (main) hsearch-quarkus 1.0-SNAPSHOT (running on Quarkus 1.0.0.CR1) started in 0.849s. Listening on: http://0.0.0.0:8080

Reviewer 2: This is with a native image? Looks weird that it doesn't start much faster. You are in prod mode right so you don't inject the data?

Author: I don't inject the data, but I initialize the schema. I believe that's why it's so slow. I guess I could write a few more scripts to initialize the DB and Elasticsearch, but I'm not sure it's reasonable to assume people will initialize the Elasticsearch schema by hand.

Strangely enough, the second run is a bit faster (0.5s) and the next ones much faster (0.1). Maybe it's related to Elasticsearch, I've noticed it takes a while to "warm up" and deliver decent response times.

Author: I added something about the warm-up.

2019-11-07 10:33:15,521 INFO  [io.quarkus] (main) Profile prod activated.
2019-11-07 10:33:15,521 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, hibernate-orm-panache, hibernate-search-elasticsearch, jdbc-postgresql, narayana-jta, resteasy, resteasy-jsonb]
----

The application started and is ready to accept commands:

[source]
----
$ curl -X PUT http://localhost:8080/client/ -H "Content-Type: application/json" -d '{"name":"Wayne Enterprises"}'

{
    "id": 1,
    "name": "Wayne Enterprises"
}
----

[source]
----
$ curl -X GET 'http://localhost:8080/client/search/?terms=enterprise'

[
    {
        "id": 1,
        "name": "Wayne Enterprises"
    }
]
----

== Beyond...

We now are the happy owners of a REST application
providing both CRUD operations
and more advanced full-text search operation,

> packaged as a container image.

Reviewer 2: You writing a haiku?

Author: A bad one, maybe :)

Because software development is a never-ending task,
there are still things we could improve:

* improving robustness by taking advantage of https://quarkus.io/guides/flyway[Flyway]

> to handle database schema uppgrade.

Reviewer 2: ```suggestion
to handle database schema upgrade.
```

> * adding https://quarkus.io/guides/native-and-ssl.html[support for SSL]

Reviewer 2: ```suggestion
* adding https://quarkus.io/guides/native-and-ssl[support for SSL]
```

Author: Right. I got on this page from the website though, so you probably have an obsolete link somewhere.

to our native binary if necessary.
* taking advantage of the many https://quarkus.io/guides/[Quarkus extensions]
to add a security layer, tracing, fault tolerance, ...
* adding more search features to our application:
** we can link:{hsearchDocUrl}#backend-elasticsearch-analysis[tune our analyzer more finely]
to get better search hits
(see https://en.wikipedia.org/wiki/Precision_and_recall[precision and recall]).
** we can link:{hsearchDocUrl}#mapper-orm-directfieldmapping[index more than just text],
including enums, numbers, date/time values, or even spatial coordinates (points).
** even custom types can be indexed thanks to link:{hsearchDocUrl}#mapper-orm-bridge[custom bridges]
** link:{hsearchDocUrl}#search-dsl-predicate[other predicates] are available,

> such as `range` ("between"/"greater than"/...), `within` (points within a circle) and many more.

Reviewer 2: I would talk about spatial predicates because it's not entirely obvious we have spatial.
